### PR TITLE
ZEPPLIN-976 ] HotFix - zeppelin server does not restart when incorrect credentials data.

### PR DESCRIPTION
**Pull request by cloverhearts**
### What is this PR 
zeppelin server does not restart when incorrect credentials data.

reproduce.
1. Click to zeppelin home for web ui.
2. Click to zeppelin Credentials.
3. 'Entity' information without writing, username and password only written to storage.
4. Click to zeppelin home for web ui.
5. Click to zeppelin Credentials.
and zeppelin restart.
##### but, Zeppelin does not work.

It creates the wrong json file. 'conf / credentials.json' according to the null.

### What type of PR is it?
Hot Fix

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-976

### How should this be tested?

1. Click to zeppelin home for web ui.
2. Click to zeppelin Credentials.
3. 'Entity' information without writing, username and password only written to storage.
4. Click to zeppelin home for web ui.
5. Click to zeppelin Credentials.
and zeppelin restart.

zeppelin does work!

### Reproduced Screenshots
zeppelin-server-error2

### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? no

### CODE REVIEW

**Comment by cloverhearts** [comment]
retry ci

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/CredentialRestApi.java [hunk 1]
```diff
@@ -63,10 +63,15 @@ public CredentialRestApi(Credentials credentials) {
...
   public Response putCredentials(String message) throws IOException {
     Map<String, String> messageMap = gson.fromJson(message,
-            new TypeToken<Map<String, String>>(){}.getType());
+      new TypeToken<Map<String, String>>(){}.getType());
     String entity = messageMap.get("entity");
     String username = messageMap.get("username");
     String password = messageMap.get("password");
+
+    if (entity == null || username == null || password == null) {
+      return new JsonResponse(Status.INTERNAL_SERVER_ERROR, "", "").build();
```
Comment (echarles) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/CredentialRestApi.java]: Status.BAD_REQUEST would be more adequate here.

Comment (cloverhearts) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/CredentialRestApi.java]: @echarles 
Thank you for your feedback.
You're right. i've changed.

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/CredentialRestApi.java [hunk 2]
```diff
@@ -63,10 +63,15 @@ public CredentialRestApi(Credentials credentials) {
   @PUT
   public Response putCredentials(String message) throws IOException {
     Map<String, String> messageMap = gson.fromJson(message,
-            new TypeToken<Map<String, String>>(){}.getType());
+      new TypeToken<Map<String, String>>(){}.getType());
     String entity = messageMap.get("entity");
     String username = messageMap.get("username");
     String password = messageMap.get("password");
+
+    if (entity == null || username == null || password == null) {
```
Comment (bzz) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/CredentialRestApi.java]: And they could not be "", or is that case is ok?

**Comment by bzz** [comment]
@cloverhearts thank you for hotfixing!

How do you think, is there a chance we could have a test that reproduces this issue under this PR or do you think it would be to hard add?

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/CredentialRestApi.java (hunk 2 above)
Comment (cloverhearts) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/CredentialRestApi.java]: An empty string does not have a problem in Zeppelin server.
I only solve a serious trouble in this pr.

**Comment by cloverhearts** [comment]
@bzz 
I think this is a very serious problem.
In my personal opinion, first merge this PR, and improve Credentials feature in future ( new pr ).

Problem situation, please refer to the screenshot.

Reproduce the problem, please note the following:
1. Click to zeppelin home for web ui.
2. Click to zeppelin Credentials.
3. 'Entity' information without writing, username and password only written to storage.
4. Click to zeppelin home for web ui.
5. Click to zeppelin Credentials.
and zeppelin restart.

**Comment by Leemoonsoo** [comment]
The fix looks good to me. But like @bzz mentioned, it will be even better if there're unittest for this problem.

**Comment by cloverhearts** [comment]
@bzz @Leemoonsoo  
I am sorry.
do you mean the test code for this feature?

**Comment by bzz** [comment]
@cloverhearts yes, if that is not very hard to do

**Comment by cloverhearts** [comment]
@bzz Okay, i would be make it.
Thank you so much.

**Comment by cloverhearts** [comment]
add test code about invalid request for credentials api

**Comment by cloverhearts** [comment]
retry ci

**Comment by minahlee** [comment]
@cloverhearts Seems like integration test fails, can you fix it please?
```
Tests run: 8, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 120.219 sec <<< FAILURE! - in org.apache.zeppelin.integration.ParagraphActionsIT
testTitleButton(org.apache.zeppelin.integration.ParagraphActionsIT)  Time elapsed: 16.504 sec  <<< FAILURE!
java.lang.AssertionError: After Hide Title : The title field contains
Expected: ""
     but: was "Untitled"
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:865)
	at org.junit.rules.ErrorCollector$1.call(ErrorCollector.java:65)
	at org.junit.rules.ErrorCollector.checkSucceeds(ErrorCollector.java:78)
	at org.junit.rules.ErrorCollector.checkThat(ErrorCollector.java:63)
	at org.apache.zeppelin.integration.ParagraphActionsIT.testTitleButton(ParagraphActionsIT.java:355)
```

**Comment by cloverhearts** [comment]
@minahlee it's not my issue.
I think the problem seems to arise from a sudden UI test master branch.
Let's solve the problem from a different PR.
One will stop working on this feature.

**Comment by astroshim** [comment]
LGTM. 
@minahlee The selenium test of travis has a problem so @corneadoug is fixing now.

**Comment by cloverhearts** [comment]
ci pass.
please, check to this pr.

**Comment by bzz** [comment]
@cloverhearts thank you for adding tests! 
Looks great to me, merging if there is no more discussion

**Pull request closed without merging**